### PR TITLE
Split 0x.js into: contract-wrappers & order-watcher

### PR DESCRIPTION
**Pull request by fabioberger**
## Description

In a previous PR, @LogvinovLeon moved `order-utils` out of 0x.js. This PR is a follow-up which moves `order-watcher` and `contract-wrappers` out of 0x.js, while keeping 0x.js's public interface intact.

Additionally, a `fill-scenarios` package was also created, which I expect to remove entirely once combinatorial testing ships. 

## How Has This Been Tested?

All the tests pass. Rendered the 0x.js docs successfully.

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [x] Change requires a change to the documentation.
*   [ ] Added tests to cover my changes.
*   [x] Added new entries to the relevant CHANGELOG.jsons.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/dev-utils/CHANGELOG.json [hunk 1]
```diff
@@ -1,4 +1,12 @@
 [
+    {
+        "version": "0.4.2",
+        "changes": [
+            {
+                "note": "Move callbackErrorReporter over from 0x.js"
```
Comment (LogvinovLeon) [inline packages/dev-utils/CHANGELOG.json]: PR number

File: packages/dev-utils/src/callback_error_reporter.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import * as chai from 'chai';
+import * as _ from 'lodash';
+
+import { DoneCallback } from '@0xproject/types';
+
+const expect = chai.expect;
+
+export const callbackErrorReporter = {
+    reportNoErrorCallbackErrors(done: DoneCallback, expectToBeCalledOnce = true) {
+        return <T>(f?: (value: T) => void) => {
```
Comment (LogvinovLeon) [inline packages/dev-utils/src/callback_error_reporter.ts]: **var-return**

File: packages/dev-utils/src/callback_error_reporter.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                    }
+                } catch (err) {
+                    done(err);
+                }
+            };
+            return wrapped;
+        };
+    },
+    reportNodeCallbackErrors(done: DoneCallback, expectToBeCalledOnce = true) {
+        return <T>(f?: (value: T) => void) => {
```
Comment (LogvinovLeon) [inline packages/dev-utils/src/callback_error_reporter.ts]: **var-return**

File: packages/fill-scenarios/CHANGELOG.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[
+    {
+        "version": "0.0.1",
+        "changes": [
+            {
+                "note": "Move FillScenarios out of 0x.js"
```
Comment (LogvinovLeon) [inline packages/fill-scenarios/CHANGELOG.json]: PR number

File: packages/fill-scenarios/package.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    "description": "0x order fill scenario generator",
+    "main": "lib/index.js",
+    "types": "lib/index.d.ts",
+    "scripts": {
+        "build:watch": "tsc -w",
+        "prebuild": "run-s clean generate_contract_wrappers",
+        "generate_contract_wrappers": "node ../abi-gen/lib/index.js --abis 'src/compact_artifacts/@(Exchange|Token|TokenTransferProxy|EtherToken).json' --template ../contract_templates/contract.handlebars --partials '../contract_templates/partials/**/*.handlebars' --output src/generated_contract_wrappers --backend ethers && prettier --write 'src/generated_contract_wrappers/**.ts'",
+        "build": "tsc && copyfiles -u 2 './src/compact_artifacts/**/*.json' ./lib/compact_artifacts && copyfiles -u 3 './lib/src/monorepo_scripts/**/*' ./scripts",
+        "clean": "shx rm -rf lib scripts",
+        "lint": "tslint --project . 'src/**/*.ts'",
```
Comment (LogvinovLeon) [inline packages/fill-scenarios/package.json]: Nit: Just `tslint --project .`

File: packages/fill-scenarios/src/index.ts [hunk 6]
```diff
@@ -1,52 +1,58 @@
...
     constructor(
-        zeroEx: ZeroEx,
+        provider: Provider,
         userAddresses: string[],
         tokens: Token[],
         zrxTokenAddress: string,
         exchangeContractAddress: string,
     ) {
-        this._zeroEx = zeroEx;
+        this._provider = provider;
```
Comment (LogvinovLeon) [inline packages/fill-scenarios/src/index.ts]: No need to store it. Get it from the `web3Wrapper` when you need it.

File: packages/fill-scenarios/src/index.ts [hunk 7]
```diff
@@ -193,11 +212,34 @@ export class FillScenarios {
...
     }
     private async _increaseAllowanceAsync(tokenAddress: string, address: string, amount: BigNumber): Promise<void> {
-        const oldMakerAllowance = await this._zeroEx.token.getProxyAllowanceAsync(tokenAddress, address);
+        const tokenInstance = new TokenContract(
+            artifacts.Token.abi,
+            tokenAddress,
+            this._provider,
+            this._web3Wrapper.getContractDefaults(),
+        );
+        const txData = {};
```
Comment (LogvinovLeon) [inline packages/fill-scenarios/src/index.ts]: Remove

File: packages/fill-scenarios/src/index.ts [hunk 8]
```diff
@@ -193,11 +212,34 @@ export class FillScenarios {
...
+        const networkArtifactsIfExists = artifacts.TokenTransferProxy.networks[constants.TEST_RPC_NETWORK_ID];
+        if (_.isUndefined(networkArtifactsIfExists)) {
+            throw new Error(`Did not find network artifacts for networkId: ${constants.TEST_RPC_NETWORK_ID}`);
+        }
+        const proxyAddress = networkArtifactsIfExists.address;
+        const oldMakerAllowance = await tokenInstance.allowance.callAsync(address, proxyAddress, txData);
```
Comment (LogvinovLeon) [inline packages/fill-scenarios/src/index.ts]: txData param has a default value

File: packages/order-utils/CHANGELOG.json [hunk 9]
```diff
@@ -1,4 +1,16 @@
 [
+    {
+        "version": "0.0.5",
+        "changes": [
+            {
+                "note": "Add formatters package for converting signedOrder to orderAddresses & orderValues",
```
Comment (LogvinovLeon) [inline packages/order-utils/CHANGELOG.json]: That is odd. You have an object with the same keys.
You should format it as a list of objects.
Also add a PR key to each one

File: packages/order-utils/src/formatters.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        ];
+        const orderValues: OrderValues = [
+            order.makerTokenAmount,
+            order.takerTokenAmount,
+            order.makerFee,
+            order.takerFee,
+            order.expirationUnixTimestampSec,
+            order.salt,
+        ];
+        return [orderAddresses, orderValues];
```
Comment (LogvinovLeon) [inline packages/order-utils/src/formatters.ts]: **var-return**

File: packages/order-watcher/CHANGELOG.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
+[
+    {
+        "version": "0.0.1",
+        "changes": [
+            {
+                "note": "Moved OrderWatcher out of 0x.js package"
```
Comment (LogvinovLeon) [inline packages/order-watcher/CHANGELOG.json]: PR number

File: packages/order-watcher/package.json [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        "exchange",
+        "orderbook"
+    ],
+    "main": "lib/src/index.js",
+    "types": "lib/src/index.d.ts",
+    "scripts": {
+        "build:watch": "tsc -w",
+        "prebuild": "run-s clean generate_contract_wrappers",
+        "generate_contract_wrappers": "node ../abi-gen/lib/index.js --abis 'src/compact_artifacts/@(Exchange|Token|TokenTransferProxy|EtherToken).json' --template ../contract_templates/contract.handlebars --partials '../contract_templates/partials/**/*.handlebars' --output src/generated_contract_wrappers --backend ethers && prettier --write 'src/generated_contract_wrappers/**.ts'",
+        "lint": "tslint --project . 'src/**/*.ts' 'test/**/*.ts'",
```
Comment (LogvinovLeon) [inline packages/order-watcher/package.json]: Same nit

File: packages/order-watcher/src/types.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+export type OnOrderStateChangeCallback = (err: Error | null, orderState?: OrderState) => void;
+
+export enum InternalOrderWatcherError {
+    NoAbiDecoder = 'NO_ABI_DECODER',
+    ZrxNotInTokenRegistry = 'ZRX_NOT_IN_TOKEN_REGISTRY',
+    WethNotInTokenRegistry = 'WETH_NOT_IN_TOKEN_REGISTRY',
+}
+
+// tslint:disable:max-file-line-count
```
Comment (LogvinovLeon) [inline packages/order-watcher/src/types.ts]: Remove

File: packages/types/CHANGELOG.json [hunk 14]
```diff
@@ -1,4 +1,13 @@
 [
+    {
+        "version": "0.7.0",
+        "changes": [
+            {
+                "note":
```
Comment (LogvinovLeon) [inline packages/types/CHANGELOG.json]: PR number

File: packages/web3-wrapper/src/web3_wrapper.ts [hunk 15]
```diff
@@ -324,6 +324,13 @@ export class Web3Wrapper {
         const txHash = await promisify<string>(this._web3.eth.sendTransaction)(txData);
         return txHash;
     }
+    /**
```
Comment (LogvinovLeon) [inline packages/web3-wrapper/src/web3_wrapper.ts]: 😍

File: packages/fill-scenarios/src/index.ts (hunk 7 above)
Comment (fabioberger) [reply packages/fill-scenarios/src/index.ts]: It is used further down. Moved it closer to where it's used.

File: packages/order-utils/src/formatters.ts (hunk 10 above)
Comment (fabioberger) [reply packages/order-utils/src/formatters.ts]: I think this is more readable then assigning it to some non-descript variable name. Don't know what I would call this.

File: packages/0x.js/src/0x.ts [hunk 16]
```diff
@@ -136,40 +131,15 @@ export class ZeroEx {
      * @param   config      The configuration object. Look up the type for the description.
      * @return  An instance of the 0x.js ZeroEx class.
      */
-    constructor(provider: Provider, config: ZeroExConfig) {
+    constructor(provider: Provider, config: ContractWrappersConfig) {
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/0x.js/src/0x.ts]: I agree that the name is not ideal but the docs do not currently support displaying types defs if we were to re-export. Think we'll leave it this way for now.

File: packages/contract-wrappers/src/contract_wrappers.ts [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+    public getProvider(): Provider {
+        return this._web3Wrapper.getProvider();
+    }
+    /*
+     * HACK: `TokenWrapper` needs a token transfer proxy address. `TokenTransferProxy` address is fetched from
+     * an `ExchangeWrapper`. `ExchangeWrapper` needs `TokenWrapper` to validate orders, creating a dependency cycle.
+     * In order to break this - we create this function here and pass it as a parameter to the `TokenWrapper`
+     * and `ProxyWrapper`.
+     */
+    private async _getTokenTransferProxyAddressAsync(): Promise<string> {
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/contract-wrappers/src/contract_wrappers.ts]: Good catch!

File: packages/0x.js/src/0x.ts [hunk 18]
```diff
@@ -178,31 +148,23 @@ export class ZeroEx {
...
+        return this._contractWrappers.getProvider();
     }
     /**
      * Get user Ethereum addresses available through the supplied web3 provider available for sending transactions.
      * @return  An array of available user Ethereum addresses.
      */
     public async getAvailableAddressesAsync(): Promise<string[]> {
-        const availableAddresses = await this._web3Wrapper.getAvailableAddressesAsync();
+        // Hack: Get Web3Wrapper from ZeroExContract
+        const web3Wrapper: Web3Wrapper = (this._contractWrappers as any)._web3Wrapper;
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/0x.js/src/0x.ts]: Agreed. Fixed this by using `getProvider()` instead and instantiating a Web3Wrapper instance here.

File: packages/contract-wrappers/test/order_validation_test.ts [hunk 19]
```diff
@@ -40,14 +40,14 @@ describe('OrderValidation', () => {
         networkId: constants.TESTRPC_NETWORK_ID,
     };
     before(async () => {
-        zeroEx = new ZeroEx(provider, config);
-        exchangeContractAddress = zeroEx.exchange.getContractAddress();
-        userAddresses = await zeroEx.getAvailableAddressesAsync();
+        contractWrappers = new ContractWrappers(provider, config);
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/contract-wrappers/test/order_validation_test.ts]: Can you elaborate?

File: packages/fill-scenarios/src/compact_artifacts/DummyToken.json [hunk 20]
```diff
@@ -0,0 +1,22 @@
+{
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/fill-scenarios/src/compact_artifacts/DummyToken.json]: Unfortunately not. Compact artifacts should like in the individual projects.

**Pull request merged**